pkg/reconciler/openshift/tektonconfig/rbac.go

```diff
@@ -393,54 +395,69 @@ func (r *rbac) createResources(ctx context.Context) error {
-		}
+		namespace := ns
+		go r.processResourcesForSingleNamespace(ctx, namespace, sem)
+	}
 
-		logger.Infow("Ensures Default SA in ", "Namespace", ns.GetName())
-		sa, err := r.ensureSA(ctx, &ns)
-		if err != nil {
-			return fmt.Errorf("failed to ensure default SA in namespace %s, %w", ns.Name, err)
-		}
+	return nil
+}
 
-		// If "operator.tekton.dev/scc" exists in the namespace, then bind
-		// that SCC to the SA
-		err = r.handleSCCInNamespace(ctx, &ns)
-		if err != nil {
-			return fmt.Errorf("failed to bind scc to namespace %s; %w", ns.Name, err)
-		}
+func (r *rbac) processResourcesForSingleNamespace(ctx context.Context, ns corev1.Namespace, sem chan struct{}) {
+	sem <- struct{}{}        // Acquire a slot in the semaphore
+	defer func() { <-sem }() // Release the slot in the semaphore
 
-		// We use a namespace scoped Role when SCC annotation is present, and
-		// a cluster scoped ClusterRole when the default SCC is used
-		roleRef := r.getSCCRoleInNamespace(&ns)
-		if err := r.ensurePipelinesSCCRoleBinding(ctx, sa, roleRef); err != nil {
-			return fmt.Errorf("failed to create Pipeline Scc Role Binding in namespace %s, %w", ns.Name, err)
-		}
+	logger := logging.FromContext(ctx)
+	logger.Infof("ensure ca bundle configmap in namespace %s", ns.Name)
+	if err := r.ensureCABundles(ctx, &ns); err != nil {
+		logger.Errorf("failed to ensure ca bundles presence in namespace %s, %w", ns.Name, err)
+	}
 
-		if err := r.ensureRoleBindings(ctx, sa); err != nil {
-			return err
-		}
+	logger.Infof("ensures default pipelines service account in namespace %s", ns.Name)
+	sa, err := r.ensureSA(ctx, &ns)
+	if err != nil {
+		logger.Errorf("failed to ensure default SA in namespace %s, %w", ns.Name, err)
+	}
 
-		if err := r.ensureClusterRoleBindings(ctx, sa); err != nil {
-			return err
-		}
+	// If "operator.tekton.dev/scc" exists in the namespace, then bind
+	// that SCC to the SA
+	err = r.handleSCCInNamespace(ctx, &ns)
+	if err != nil {
+		logger.Errorf("failed to bind scc to namespace %s; %w", ns.Name, err)
+	}
+
+	// We use a namespace scoped Role when SCC annotation is present, and
+	// a cluster scoped ClusterRole when the default SCC is used
+	roleRef := r.getSCCRoleInNamespace(&ns)
+	if err := r.ensurePipelinesSCCRoleBinding(ctx, sa, roleRef); err != nil {
+		logger.Errorf("failed to create Pipeline Scc Role Binding in namespace %s, %w", ns.Name, err)
+	}
 
-		// Add `openshift-pipelines.tekton.dev/namespace-reconcile-version` label to namespace
-		// so that rbac won't loop on it again
+	if err := r.ensureRoleBindings(ctx, sa); err != nil {
+		logger.Errorf("failed to create rolebinding in namespace %s, %w", ns.Name, err)
+	}
+
+	if err := r.ensureClusterRoleBindings(ctx, sa); err != nil {
+		logger.Errorf("failed to create clusterrolebinding in namespace %s, %w", ns.Name, err)
+	}
+
+	// if No error add `openshift-pipelines.tekton.dev/namespace-reconcile-version` label to namespace
+	// so that rbac won't loop on it again
+	if err == nil {
```

Comment:
err will be nil right if above function failed but last one succeed?

Reply:
Yeah, thank you for catching this @piyush-garg, proposing a fix

Reply:
seems that might be a data race using the shared logger stored in the context with 
logger := logging.FromContext(ctx), it returns a zaplogger that is shared between goroutines and fortunately the issue happened on one of the tests
https://storage.googleapis.com/tekton-prow/pr-logs/pull/tektoncd_operator/2218/pull-tekton-operator-unit-tests/1809201707154935808/build-log.txt

`WARNING: DATA RACE
Read at 0x00c00051b563 by goroutine 45:
  testing.(*common).logDepth()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/testing/testing.go:1017 +0xcd
  testing.(*common).log()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/testing/testing.go:1010 +0xa4
  testing.(*common).Logf()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/testing/testing.go:1061 +0x6a
  testing.(*T).Logf()
      <autogenerated>:1 +0x69
  go.uber.org/zap/zaptest.TestingWriter.Write()
      /home/prow/go/src/github.com/tektoncd/operator/vendor/go.uber.org/zap/zaptest/logger.go:146 +0x11d
  go.uber.org/zap/zaptest.(*TestingWriter).Write()
      <autogenerated>:1 +0x74
  go.uber.org/zap/zapcore.(*ioCore).Write()
      /home/prow/go/src/github.com/tektoncd/operator/vendor/go.uber.org/zap/zapcore/core.go:99 +0x192
  go.uber.org/zap/zapcore.(*CheckedEntry).Write()
      /home/prow/go/src/github.com/tektoncd/operator/vendor/go.uber.org/zap/zapcore/entry.go:253 +0x1ef
  go.uber.org/zap.(*SugaredLogger).log()
      /home/prow/go/src/github.com/tektoncd/operator/vendor/go.uber.org/zap/sugar.go:355 +0x12f
  go.uber.org/zap.(*SugaredLogger).Infof()
      /home/prow/go/src/github.com/tektoncd/operator/vendor/go.uber.org/zap/sugar.go:204 +0x244
  github.com/tektoncd/operator/pkg/reconciler/openshift/tektonconfig.(*rbac).processResourcesForSingleNamespace()
      /home/prow/go/src/github.com/tektoncd/operator/pkg/reconciler/openshift/tektonconfig/rbac.go:414 +0x19a
  github.com/tektoncd/operator/pkg/reconciler/openshift/tektonconfig.(*rbac).createResources.gowrap1()
      /home/prow/go/src/github.com/tektoncd/operator/pkg/reconciler/openshift/tektonconfig/rbac.go:401 +0x124

Previous write at 0x00c00051b563 by goroutine 42:
  testing.tRunner.func1()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/testing/testing.go:1676 +0x8fa
  runtime.deferreturn()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/runtime/panic.go:602 +0x5d
  testing.(*T).Run.gowrap1()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/testing/testing.go:1742 +0x44

Goroutine 45 (running) created at:
  github.com/tektoncd/operator/pkg/reconciler/openshift/tektonconfig.(*rbac).createResources()
      /home/prow/go/src/github.com/tektoncd/operator/pkg/reconciler/openshift/tektonconfig/rbac.go:401 +0x3e4
  github.com/tektoncd/operator/pkg/reconciler/openshift/tektonconfig.TestCreateResources.func1()
      /home/prow/go/src/github.com/tektoncd/operator/pkg/reconciler/openshift/tektonconfig/rbac_test.go:175 +0x5e7
  testing.tRunner()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/testing/testing.go:1742 +0x44

Goroutine 42 (finished) created at:
  testing.(*T).Run()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/testing/testing.go:1742 +0x825
  testing.runTests.func1()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/testing/testing.go:2161 +0x85
  testing.tRunner()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/testing/testing.go:1689 +0x21e
  testing.runTests()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      /home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.linux-amd64/src/testing/testing.go:2027 +0xf17
  main.main()
      _testmain.go:51 +0x2bd`